src/System.Linq.Expressions/tests/ILReader/DynamicMethodILProvider.cs

```diff
@@ -0,0 +1,50 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+// Code adapted from https://blogs.msdn.microsoft.com/haibo_luo/2010/04/19/ilvisualizer-2010-solution
+
+using System.Reflection;
+using System.Reflection.Emit;
+
+namespace System.Linq.Expressions.Tests
+{
+    public sealed class DynamicMethodILProvider : IILProvider
+    {
+        private static readonly FieldInfo s_fiLen = typeof(ILGenerator).GetField("m_length", BindingFlags.NonPublic | BindingFlags.Instance);
+        private static readonly FieldInfo s_fiStream = typeof(ILGenerator).GetField("m_ILStream", BindingFlags.NonPublic | BindingFlags.Instance);
+        private static readonly MethodInfo s_miBakeByteArray = typeof(ILGenerator).GetMethod("BakeByteArray", BindingFlags.NonPublic | BindingFlags.Instance);
+
+        private readonly DynamicMethod m_method;
+        private byte[] m_byteArray;
+
+        public DynamicMethodILProvider(DynamicMethod method)
+        {
```

Comment:
Nit: consider adding asserts that the various Field/MethodInfos captured in the static ctor are non-null.  If those field names ever change, it'll be easier for someone to spot the problem if an assert triggers rather than getting null ref exceptions while running tests.